sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -661,6 +690,26 @@ case class HashShuffleSpec(
   override def numPartitions: Int = partitioning.numPartitions
 }
 
+case class CoalescedHashShuffleSpec(
+    from: ShuffleSpec,
+    partitions: Seq[CoalescedBoundary]) extends ShuffleSpec {
+
+  override def isCompatibleWith(other: ShuffleSpec): Boolean = other match {
+    case SinglePartitionShuffleSpec =>
+      numPartitions == 1
+    case CoalescedHashShuffleSpec(otherParent, otherPartitions) =>
+      partitions == otherPartitions && from.isCompatibleWith(otherParent)
```

Comment:
suppose both `from` and `otherParent` are `HashShuffleSpec`, is it possible that they may have different number of partitions, but after coalescing the number of partitions become the same? 

In this case we'll consider the two incompatible but they actually should be?

Reply:
Hmm, this is not related to my above comment. The check `partitions == otherPartitions` is OK, my question is on `from.isCompatibleWith(otherParent)`, which checks (assuming both are `HashShuffleSpec`) whether the original number of partitions are the same.

Reply:
I think it must to be `HashShuffleSpec`. You mean that their partition numbers can be different but compatible?

Reply:
I mean their partition numbers are different and thus `from.isCompatibleWith(otherParent)` will return false, and thus cause this `CoalescedHashShuffleSpec.isCompatibleWith` to also return false. But should we return true instead if the partitions after coalescing are the same?

Reply:
In other words, `from.isCompatibleWith(otherParent)` could be too conservative

Reply:
Hmm, if first hash partitioning is 5 partitions, second is 4 partitions. How can we get same coalesced partitions with that?

For example:

[[0, 3], [3, 5]] != [[0, 3], [3, 4]]
[[0, 2], [2, 3], [3, 5]] != [[0, 2], [2, 3], [3, 4]]

The end reducer index of last coalesced partition should be different always, no?

Reply:
I'm not sure whether it is possible for this case to happen. But irrespective of that, I feel this check is unnecessary here.

Reply:
But what about two (nonsensical) CoalescedHashShuffleSpec where is partition is just coalesced from single partition from the some parent HashShuffleSpec. Is it not then correct to do the `from.isCompatibleWith(otherParent)` check? or is it expected that we can make them compatible by just coalescing them in such a way?